### PR TITLE
Doc: Initial version of ARCN RTVM Perf Tips

### PR DESCRIPTION
**Pull request by deb-intel**
Signed-off-by: Deb Taylor <deb.taylor@intel.com>

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: doc/tutorials/rtvm_performance_tips.rst [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+##################################
+
+Background
+**********
+
+The ACRN real-time VM (RTVM) is a special type of ACRN post-launched VM. In
+order to achieve bare metal-like RT performance, a set of constraints and
+technologies are applied to the RTVM compared to the ACRN standard VM. With
+these additional constraints and technologies, RT tasks can run on the RTVM
+without a VM-exit, which is a key virtualization overhead issue.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:13]: I think a better phrasing would be: "RT tasks can run in an RTVM without incurring `VM-exit` instructions. Those are a primary source of performance overhead when running in a virtualized environment". (or something similar)

File: doc/tutorials/rtvm_performance_tips.rst [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+**Tip 1:** Apply the acrn-dm option "--lapic_pt" and make the guest RTVM
+operate under the LAPIC X2APIC mode to enable the LAPIC passthrough.
+
+The LAPIC passthrough feature of ACRN is configured via the "--lapic_pt"
+option, but the feature is actually enabled when LAPIC is switched to X2APIC
+mode. So, both conditions should be met to enable an RTVM. The "--rtvm"
+option will be automatically attached once "--lapic_pt" is applied.
+
+**Tip 2:** If necessary, use virtio polling mode to prevent the frontend of
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:51]: I think the idea is that it's the frontend part of the virtio driver (i.e. the part running in the User VM) that may trigger a `VM-Exit` if polling mode is **not** used. Can we find a better formulation for this here?

File: doc/tutorials/rtvm_performance_tips.rst [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+the VM-exit from sending a notification to the backend.
+
+We recommend that you passthrough a physical peripheral device to an RTVM,
+such as block or an ethernet device. If no physical device is available,
+ACRN supports virtio devices and enables the polling mode to avoid a VM-exit
+at the frontend. Virtio polling mode can be enabled via the option
+"--virtio_poll [polling interval]".
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:58]: Same here about the overall formulation.

Also, a nitpick: use double-backticks around `--virtio_poll <polling interval>` (and use greater than and smaller than characters to be consistent with the `acrn-dm` user guide)

File: doc/tutorials/rtvm_performance_tips.rst [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+VM-exit has a significant negative impact on virtualization performance.
+A single VM-exit can cause several micro-second latencies, or even longer,
+depending on what's done in VMX-root mode. VM-exit is classified into two
+types: triggered by external CPU events or triggered by operations initiated
+by the vCPU.
+
+ACRN eliminates almost all VM-exits triggered by external events via the
+LAPIC passthrough. A few exceptions exist:
+
+- SMI - it will bring the processor into the SMM, causing a much longer
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:72]: Can we spell out the acronym (at least once) - System Management Interrupt == SMI

File: doc/tutorials/rtvm_performance_tips.rst [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  performance impact. The SMI should be handled in the BIOS.
+
+- NMI - ACRN uses NMI for system-level notification.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:75]: Can we spell out the acronym (at least once) - Non-Maskable Interrupt == NMI

File: doc/tutorials/rtvm_performance_tips.rst [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+Users should take care of VM-exits that are triggered by operations
+initiated by the vCPU. Refer to the Intel SMD: "Instructions Cause VM-exits
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:78]: SMD -> SDM

File: doc/tutorials/rtvm_performance_tips.rst [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+Unconditionally" (SDM V3, 25.1.2) and "Instructions That Cause VM-exits
+Conditionally" (SDM V3, 25.1.3).
+
+**Tip 3:** Do not use CPUID in the RT critical section.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:82]: Nitpick: I would enclose all `CPUID`, `RDTSCP`, `RDTSC`, `RDMSR` and `WRMSR` in double-backticks

File: doc/tutorials/rtvm_performance_tips.rst [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+application, and a typical case is using CPUID immediately before and after
+RDTSC. In order to remove CPUID in this case, use RDTSCP instead of RDTSC.
+Because RDTSCP waits until all previous instructions have been executed
+before reading the counter, and the subsequent instructions after the RDTSCP
+normally have data dependency on it, they must wait until the RDTSCP has
+been executed.
+
+RDMSR or WRMSR are instructions that cause VM-exits conditionally. On the
+ACRN RTVM, most MSRs are not intercepted by the HV, so they won't cause a
+VM-exit. But there are exceptions for security consideration: 1) read from
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:98]: Should we make these a numbered list?

File: doc/tutorials/rtvm_performance_tips.rst [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+ACRN uses VMX_TSC_OFFSET_FULL as the offset between vTSC_ADJUST and
+pTSC_ADJUST; therefore, if VMX_TSC_OFFSET_FULL is zero, intercepting
+TSC_ADJUST and TSC_DEADLINE is not necessary. Otherwise, they should be
+intercepted to guarantee functionality.
+
+**Tip 6:** Utilize Preempt-RT Linux mechanisms to reduce the access of ICR
+from the RT core:
+
+#. Add "domain" to the "isolcpus" ( “isolcpus=nohz,domain,1” ) to the kernel parameters.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:120]: Use of double-backticks instead of quotes?

File: doc/tutorials/rtvm_performance_tips.rst [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+#. Add "idle=poll" to the kernel parameters.
+#. Add "rcu_nocb_poll" along with "rcu_nocbs=1" to the kernel parameters.
+#. Disable the logging service like journald, syslogd if possible.
+
+These parameters are recommended for the guest Preempt-RT Linux. For a UP
+RTVM, ICR interception is not a problem. But for an SMP RTVM, IPI may be
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:126]: Can we spell out IP once (Intel-Process Interrupt) and ICR (Interrupt Command Register)

File: doc/tutorials/rtvm_performance_tips.rst [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+However, for RT VM, CPUs should be dedicatedly allocated for the determinism.
+
+**TIP 9:** Use RDT such as CAT and MBA to allocate dedicated resources to
+the RTVM.
+
+ACRN enables the Intel® Resource Director Technology, such as CAT and MBA,
+components such as the GPU via memory hierarchy. The availability of RDT is
+hardware-specific. Refer to the :ref:`rdt_configuration`.
+
+**TIP 10:** Lock the GPU to a feasible lowest frequency.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:166]: What does "a feasible lowest" mean? I guess the intention is to indicate that one should lower the GPU frequency to the lowest frequency to guarantees proper operations of the other VMs in the system that use it?

File: doc/tutorials/rtvm_performance_tips.rst [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+GPU can put heavy pressure on the power/memory subsystem, so locking the GPU
+frequency as low as possible can help to improve the determinism of RT
+performance. It can be locked in the BIOS, but the availability of certain
+BIOS option is platform-specific.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:171]: I believe there are user-space tools that can be used in the Service VM to do this as well, should the Bios not support it (`intel_gpu_frequency`, and possibly by directly tweaking some files in `/sys` too?). Can we add this?

File: doc/tutorials/rtvm_performance_tips.rst [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+Miscellaneous
+*************
+
+**TIP 11:**  Disable timer migration on Preempt-RT Linux.
+
+Because most tasks are set affinitive to the housekeeping core, the timer
+armed by RT tasks might be migrated to the nearest busy CPU for power
+saving. But it will hurt the determinism because the timer interrupts raised
+on the housekeeping core need to be resent to the RT core. The timer
+migration could be disabled by cmd: "echo 0 > /proc/kernel/timer_migration"
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:182]: Nitpick: use double-backticks instead

File: doc/tutorials/rtvm_performance_tips.rst [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+
+**TIP 12:** Add "mce=off" to RT VM kernel parameters.
```
Comment (gvancuts) [inline doc/tutorials/rtvm_performance_tips.rst:184]: Nitpick: use double-backticks instead

**Pull request merged**